tasks_test.go

```diff
@@ -0,0 +1,200 @@
+package druid
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/gocarina/gocsv"
+	"github.com/google/uuid"
+	"github.com/jmoiron/sqlx/types"
+	"github.com/stretchr/testify/require"
+	tc "github.com/testcontainers/testcontainers-go/modules/compose"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+// testDO represents entry with payload.
+type testDO struct {
+	Timestamp time.Time      `db:"ts"`
+	Id        uuid.UUID      `db:"id"`
+	Payload   types.JSONText `db:"payload"`
+}
+
+var testObjects = []testDO{
+	{
+		Id:        uuid.New(),
+		Timestamp: time.Now(),
+		Payload:   types.JSONText("{\"test\": \"json\"}"),
+	},
+	{
+		Id:        uuid.New(),
+		Timestamp: time.Now().Add(time.Hour),
+		Payload:   types.JSONText("{\"test\": \"json2\"}"),
+	},
+}
+
+func triggerIngestionTask(d *Client, dataSourceName string, entries []testDO) (string, error) {
+	csvEntriesBuff := &bytes.Buffer{}
+
+	err := gocsv.MarshalWithoutHeaders(entries, csvEntriesBuff)
+	if err != nil {
+		return "", err
+	}
+
+	var spec = NewTaskIngestionSpec(
+		SetTaskType("index_parallel"),
+		SetTaskDataSource(dataSourceName),
+		SetTaskTuningConfig("index_parallel", 25000, 5000000),
+		SetTaskIOConfigType("index_parallel"),
+		SetTaskInputFormat("csv", "false", []string{"ts", "id", "payload"}),
+		SetTaskInlineInputData(csvEntriesBuff.String()),
+	)
+	taskID, err := d.Tasks().SubmitTask(spec)
+	return taskID, err
+}
+
+type testFunction func() error
+
+func awaitTaskCompletion(client *Client, taskID string, durationSeconds int, tickDurationMilliseconds time.Duration) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		time.Sleep(time.Duration(durationSeconds) * time.Second)
+		cancel()
+	}()
+	ticker := time.NewTicker(tickDurationMilliseconds * time.Millisecond)
+	for {
+		select {
+		case <-ticker.C:
+			res, err := client.Tasks().GetStatus(taskID)
+			if err != nil {
+				return err
+			}
+
+			if res.Status.Status == "RUNNING" {
+				continue
+			}
+			break
+		case <-ctx.Done():
+			return errors.New("awaitTaskCompletion timeout")
+		}
+	}
+}
+
+func awaitTaskRunning(client *Client, taskID string, durationSeconds int, tickDurationMilliseconds time.Duration) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		time.Sleep(time.Duration(durationSeconds) * time.Second)
+		cancel()
+	}()
+	ticker := time.NewTicker(tickDurationMilliseconds * time.Millisecond)
+	for {
+		select {
+		case <-ticker.C:
+			res, err := client.Tasks().GetStatus(taskID)
+			if err != nil {
+				return err
+			}
+
+			if res.Status.Status == "RUNNING" {
+				return nil
+			}
+		case <-ctx.Done():
+			return errors.New("awaitTaskCompletion timeout")
+		}
+	}
+}
+
+func runInlineIngestionTask(client *Client, dataSourceName string, entries []testDO, recordsCount int) error {
```

Comment:
as we're exposing tasks API, we might as well expose those three methods from this test (and then use them in telemetry repo) - not necessarily in this PR though.

WDYT?

Reply:
Thank you for the suggestion. I switched to template functions to accept different input data and exposed the template functions.